// rapydscript/amd/test.js
'use strict'
if (typeof define !== 'function') { var define = require('amdefine')(module) }

define([
"path",
"fs",
"assert",
"rapydscript/test",
"rapydscript/build",
"../../../tools/node"
],

function (path, fs, assert, test, build, RapydScript) {

    function _$rapyd$_extends(child, parent) {
        child.prototype = Object.create(parent.prototype);
        child.prototype.constructor = child;
    }

    function Tests() {
        test.Tests.prototype.constructor.apply(this, arguments);
        this.options.amd = true;
    }

    _$rapyd$_extends(Tests, test.Tests);

    Tests.prototype.set_up = function set_up() {
        console.log("Calling sublclass setup");
        test.Tests.prototype.set_up.apply(this, arguments);

        this.options.node_modules = true;

        // Setup a node_modules directory in the testing dir
        this.node_modules_dir = path.join(this.test_dir, "node_modules");
        fs.mkdir(this.node_modules_dir);
    }

    Tests.prototype.compile_imports = function compile_imports() {
        this.builder.compile_imports();
    }

    Tests.prototype.run_test_case = function run_test_case() {
        var builder = this.builder;
        var test_runner ="exports.run = function(){" +
                          "require('" + builder.module_name() + "');" +
                          "};";

        // The actual test code.
        builder.write();

        // Run the test code through the test runner.
        var testcase_filepath = builder.js_filepath()+"test.js";
        fs.writeFileSync(testcase_filepath, test_runner);
        var testcase = require(testcase_filepath);
        testcase.run();
    }

    return {
        Tests : Tests
    };

});
